git/branch_manager.go

```diff
@@ -0,0 +1,96 @@
+package git
+
+import (
+	"fmt"
+	"regexp"
+
+	"github.com/drewwyatt/git-tidy/reporting"
+)
+
+type branchDeletionError struct {
+	Branch string
+	Msg    string
+}
+
+var reporter = reporting.NewReporter()
+
+// BranchManager higher level abstraction of Git
+type BranchManager struct {
+	git Git
+
+	GoneBranches         []string
+	DeletedBranches      []string
+	BranchDeletionErrors []branchDeletionError
+}
+
+// NewBranchManager Constructor for branch manager
+func NewBranchManager(directory string, exec Executor) BranchManager {
+	git := Git{directory: directory, exec: exec}
+	return BranchManager{git: git}
+}
+
+// Init run setup git commands and set gone branches
+func (bm *BranchManager) Init() []error {
+	var possibleErrors []error
+	possibleErrors = append(possibleErrors, bm.fetch())
+	possibleErrors = append(possibleErrors, bm.prune())
+	possibleErrors = append(possibleErrors, bm.listRemoteBranches())
+
+	var errors []error
+	for idx, error := range possibleErrors {
+		if error != nil {
+			errors = append(errors, error)
+			reporter.FailProcess(idx, error.Error())
+		} else {
+			reporter.PassProcess(idx)
+		}
+	}
+
+	if len(errors) == 0 {
+		bm.GoneBranches = bm.findGoneBranches()
+		reporter.Complete("Initialization complete")
+	} else {
+		reporter.Fail("Failed to initialize")
+	}
+
+	return errors
+}
+
+func (bm *BranchManager) fetch() error {
```

Comment:
I know we've hashed this out already somewhere, but it feels weird that all these are methods on `*BranchManager` but you're calling them on a `BranchManager`

Reply:
I think in the previous case, where we were returning the passed in pointer, we needed the star. Here, i am having trouble telling a difference. What makes you say:

> but you're calling them on a BranchManager

Reply:
I mean `NewBranchManager` returns a value, not a pointer, and you're not explicitly creating a pointer.

Reply:
> you're not explicitly creating a pointer.

In Go, you don't have to explicitly pass a pointer to a pointer receiver, it's done for you ([you can see an example of that here](https://tour.golang.org/methods/4)). 

I still don't really have a strong preference, but the docs on that page do indicate: 

> Since methods often need to modify their receiver, pointer receivers are more common than value receivers.

at least some of the methods (e.g. `Delete`) on `BranchManager` *do* modify the receiver. I *could* update the methods that don't modify the receiver to be value receivers, but that feels a little odd to me - though I can't quite explain why.

Reply:
I'm not suggesting changing the receivers to value receivers. I'm suggesting changing `NewBranchManager` to return a pointer.

Reply:
Right, but `NewBranchManager` doesn't **need** to return a pointer. I can't find an official example of how Go wants you to write constructors, but I did find a [better example of the point I was trying to make in my previous comment](https://tour.golang.org/methods/6).

Specifically:

> For the statement v.Scale(5), even though v is a value and not a pointer, the method with the pointer receiver is called automatically. That is, as a convenience, Go interprets the statement v.Scale(5) as (&v).Scale(5) since the Scale method has a pointer receiver.